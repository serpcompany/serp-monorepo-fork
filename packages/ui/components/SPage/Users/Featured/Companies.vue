<script setup lang="ts">
  import type { Company, Category } from '@serp/types/types';

  const { loggedIn, user } = useUserSession();

  const loading = ref(false);

  const toast = useToast();

  const form = reactive({
    category: '',
    company: '',
    placement: ''
  });

  let data = [];
  if (loggedIn.value) {
    data = await useCompanyFeaturedSubscriptions(false);
  }

  let categories = await useCompanyCategories();
  if (!categories) {
    categories = [];
  }
  const categoryOptions = ref(categories?.map((category) => category.name));

  const submitLoading = ref(false);
  const disabled = ref(true);
  const companies = ref([]);
  const companyOptions = ref(companies.value.map((company) => company.slug));
  const availablePlacements = ref([]);
  async function getCompaniesForCategory(category: string) {
    try {
      const categorySlug = categories.find(
        (cat) => cat.name === category
      )?.slug;

      const companiesData = await useAllCompaniesForCategory(categorySlug);
      if (companiesData && companiesData.entities?.length) {
        companies.value = companiesData.entities;
        companyOptions.value = companiesData.entities.map(
          (company: Company) => company.slug
        );
        availablePlacements.value = companiesData.availablePlacements;
        disabled.value = false;
      } else {
        companies.value = [];
        companyOptions.value = [];
        availablePlacements.value = [];
        disabled.value = true;
      }
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error('Error fetching companies:', error);
    }
  }

  function getCompanyId(company: string) {
    const companyData = companies.value.find((c) => c.slug === company);
    return companyData ? companyData.id : null;
  }

  function getCategoryId(category: string) {
    const categoryData = categories.find((c) => c.name === category);
    return categoryData ? categoryData.id : null;
  }

  function getCategorySlug(category: string) {
    const categoryData = categories.find((c) => c.name === category);
    return categoryData ? categoryData.slug : null;
  }

  async function submitFeaturedCompany() {
    try {
      if (!loggedIn.value) {
        toast.add({
          id: 'not-logged-in',
          title: 'Not Logged In',
          description: 'Please log in to feature a company.',
          icon: 'exclamation-circle'
        });
        return;
      }
      if (!form.category || !form.company || !form.placement) {
        toast.add({
          id: 'missing-fields',
          title: 'Missing Fields',
          description: 'Please fill in all fields.',
          icon: 'exclamation-circle'
        });
        return;
      }

      submitLoading.value = true;
      // Check if placement is still open
      const data = await useCheckPlacementAvailability(
        form.placement,
        getCompanyId(form.company),
        form.category === 'all' ? null : getCategorySlug(form.category)
      );
      if (data.reservationId) {
        // If placement is available, proceed with payment
        const successRoute =
          '/users/purchase?success=true&redirectTo=/users/get-featured';
        const cancelRoute =
          '/users/purchase?cancel=true&redirectTo=/users/get-featured';
        const response = await $fetch(
          `/api/stripe/create-checkout-session?type=featured-${form.placement}&id=${getCompanyId(
            form.company
          )}&secondaryId=${getCategoryId(form.category)}&successRoute=${encodeURIComponent(successRoute)}&cancelRoute=${encodeURIComponent(cancelRoute)}`,
          {
            method: 'GET'
          }
        );

        if (response) {
          window.location.href = response;
        }
      } else {
        toast.add({
          id: 'placement-not-available',
          title: 'Placement Not Available',
          description: 'This placement is no longer available.',
          icon: 'exclamation-circle'
        });
      }
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error('Error submitting featured company:', error);
      toast.add({
        id: 'submission-error',
        title: 'Submission Error',
        description: 'An error occurred while submitting the featured company.',
        icon: 'exclamation-circle'
      });
    } finally {
      submitLoading.value = false;
    }
  }

  useSeoMeta({
    title: 'Featured'
  });
</script>

<template>
  <UPage>
    <UMain>
      <UContainer class="mx-auto mt-16 max-w-3xl">
        <div v-if="loggedIn">
          <!-- Featured Companies List -->
          <div v-if="data.length" class="mb-12">
            <h2 class="mb-6 text-2xl font-bold">Your Featured Companies</h2>
            <div class="grid grid-cols-1 gap-4 md:grid-cols-2 lg:grid-cols-3">
              <UCard
                v-for="subscription in data"
                :key="subscription.id"
                class="transition-all hover:shadow-lg"
              >
                <template #header>
                  <div class="flex items-center justify-between">
                    <h3 class="text-lg font-semibold">
                      {{ subscription.entityName }}
                    </h3>
                    <UBadge
                      v-if="subscription.isActive"
                      color="success"
                      variant="soft"
                      >Active</UBadge
                    >
                  </div>
                </template>

                <div class="space-y-2">
                  <NuxtLink
                    :to="
                      subscription.categorySlug
                        ? `/products/best/${subscription.categorySlug}/`
                        : `/products`
                    "
                    class="text-gray-600 hover:underline"
                  >
                    Category:
                    {{
                      subscription.categoryName
                        ? subscription.categoryName
                        : 'All'
                    }}
                  </NuxtLink>
                  <p class="text-gray-600">
                    Placement: {{ subscription.placement }}
                  </p>
                  <p class="text-sm text-gray-500">
                    Created: {{ subscription.createdAt }}
                  </p>
                </div>

                <template #footer>
                  <NuxtLink
                    :to="`/products/${subscription.entitySlug}/reviews/`"
                    class="text-primary-600 font-medium hover:underline"
                  >
                    View Company â†’
                  </NuxtLink>
                </template>
              </UCard>
            </div>
          </div>

          <!-- Feature a Company Form -->
          <UCard>
            <template #header>
              <div class="flex items-center justify-between">
                <h2 class="text-xl font-semibold">Get Featured</h2>
                <UBadge color="info" variant="soft"
                  >Promote Your Company</UBadge
                >
              </div>
            </template>

            <div class="space-y-5 py-4">
              <p class="mb-4 text-gray-500 dark:text-gray-400">
                Premium placement on pages that drive leads for your brand.
              </p>

              <UFormField
                label="Category"
                help="Select a category for your featured listing"
                class="w-full"
              >
                <UInputMenu
                  v-model="form.category"
                  :items="categoryOptions"
                  class="w-full"
                  placeholder="Select a category"
                  @change="getCompaniesForCategory(form.category)"
                />
              </UFormField>

              <UFormField label="Company" class="w-full">
                <UInputMenu
                  v-model="form.company"
                  :items="companyOptions"
                  :disabled="disabled"
                  class="w-full"
                  placeholder="Select company"
                />
              </UFormField>

              <UFormField label="Position" class="w-full">
                <UInputMenu
                  v-model="form.placement"
                  :items="availablePlacements"
                  :disabled="disabled"
                  class="w-full"
                  placeholder="Position you want your company featured in"
                />
              </UFormField>
            </div>

            <template #footer>
              <div class="flex justify-end">
                <UButton
                  color="primary"
                  variant="solid"
                  size="xl"
                  :loading="submitLoading"
                  :disabled="
                    disabled ||
                    !form.category ||
                    !form.company ||
                    !form.placement
                  "
                  @click="submitFeaturedCompany"
                >
                  Submit
                </UButton>
              </div>
            </template>
          </UCard>
        </div>

        <!-- Not Logged In View -->
        <div v-else>
          <UCard class="py-8 text-center">
            <h2 class="mb-4 text-xl font-semibold">Get Featured</h2>
            <p class="mb-6 text-gray-500 dark:text-gray-400">
              Login to feature your company on our platform for increased
              visibility.
            </p>
            <UButton to="/login" color="primary"> Login to Continue </UButton>
          </UCard>
        </div>
      </UContainer>
    </UMain>
  </UPage>
</template>
