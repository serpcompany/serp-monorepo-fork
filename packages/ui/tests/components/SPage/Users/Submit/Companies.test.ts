import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { describe, expect, it, vi } from 'vitest';
import { ref } from 'vue';
import ComponentRender from '../../../../componentRender';
import Companies from '../../../../../components/SPage/Users/Submit/Companies.vue';

// Set up global mocks for session and data fetching.
let session_: { loggedIn: unknown; user: unknown; clear: unknown };
let companySubmissions_: unknown = null;

// Mock all Nuxt imports
mockNuxtImport('useUserSession', () => () => session_);
mockNuxtImport('useCompanySubmissions', () => async () => companySubmissions_);
mockNuxtImport('navigateTo', () => () => {});
mockNuxtImport('useSeoMeta', () => () => {});

describe('SPage/Users/Submit/Companies Snapshot', () => {
  const scenarios: [string, { session: unknown; companyData: unknown }][] = [
    [
      'with submissions',
      {
        session: {
          loggedIn: ref(true),
          user: ref({ email: 'test@example.com', name: 'Test User' }),
          clear: vi.fn()
        },
        companyData: [
          {
            id: 1,
            isPriority: true,
            approved: true,
            domain: 'example.com',
            logo: 'https://example.com/logo.png',
            name: 'Example Company',
            description: 'A sample description',
            reviewedAt: '2023-01-01',
            reviewedNotes: 'Looks good',
            reviewedBy: 'admin@example.com',
            createdAt: '2022-12-01'
          },
          {
            id: 2,
            isPriority: false,
            approved: false,
            domain: 'test.com',
            logo: null,
            name: 'Test Company',
            description: '',
            reviewedAt: null,
            reviewedNotes: null,
            reviewedBy: null,
            createdAt: '2022-11-15'
          }
        ]
      }
    ],
    [
      'with no submissions',
      {
        session: {
          loggedIn: ref(true),
          user: ref({ email: 'test@example.com', name: 'Test User' }),
          clear: vi.fn()
        },
        companyData: []
      }
    ]
  ];

  it.each(scenarios)(
    'renders %s correctly',
    async (desc: string, { session, companyData }) => {
      // Set the current session and mock submission data.
      session_ = session;
      companySubmissions_ = companyData;

      const html = await ComponentRender(
        `SPage/Users/Submit/Companies ${desc}`,
        {},
        Companies
      );
      expect(html).toMatchSnapshot();
    }
  );
});
