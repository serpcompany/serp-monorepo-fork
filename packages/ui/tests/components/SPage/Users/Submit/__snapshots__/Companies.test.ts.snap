// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SPage/Users/Submit/Companies Snapshot > renders with no submissions correctly 1`] = `
"<div class="flex flex-col lg:grid lg:grid-cols-10 lg:gap-10">
  <!--v-if-->
  <div class="lg:col-span-10">
    <!-- hero -->
    <div>
      <div class="py-20">
        <div class="mx-auto max-w-5xl px-4 sm:px-6 lg:px-8">
          <div class="items-center">
            <div class="flex flex-col items-center text-center lg:items-start lg:text-left">
              <h1 class="my-4 text-2xl font-bold text-pretty lg:text-5xl">Submissions</h1>
              <!--v-if-->
              <div class="flex w-full flex-col justify-center gap-2 sm:flex-row lg:justify-start">
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <main class="min-h-[calc(100vh-var(--ui-header-height))]">
      <!-- rows: companies -->
      <div class="space-y-4">
        <div class="grid grid-cols-1 gap-4 md:grid-cols-2 lg:grid-cols-3"></div>
      </div>
    </main>
  </div>
  <!--v-if-->
</div>"
`;

exports[`SPage/Users/Submit/Companies Snapshot > renders with submissions correctly 1`] = `
"<div class="flex flex-col lg:grid lg:grid-cols-10 lg:gap-10">
  <!--v-if-->
  <div class="lg:col-span-10">
    <!-- hero -->
    <div>
      <div class="py-20">
        <div class="mx-auto max-w-5xl px-4 sm:px-6 lg:px-8">
          <div class="items-center">
            <div class="flex flex-col items-center text-center lg:items-start lg:text-left">
              <h1 class="my-4 text-2xl font-bold text-pretty lg:text-5xl">Submissions</h1>
              <!--v-if-->
              <div class="flex w-full flex-col justify-center gap-2 sm:flex-row lg:justify-start">
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <main class="min-h-[calc(100vh-var(--ui-header-height))]">
      <!-- rows: companies -->
      <div class="space-y-4">
        <div class="grid grid-cols-1 gap-4 md:grid-cols-2 lg:grid-cols-3">
          <div class="rounded-lg bg-white p-4 shadow-md dark:bg-neutral-800">
            <div><span class="font-medium inline-flex items-center text-xs px-2 py-1 gap-1 rounded-md bg-primary text-inverted"><!--v-if-->Priority<!--v-if--></span></div><a href="/products/example.com/reviews/"><img alt="Submission Image" data-nuxt-img="" srcset="https://example.com/logo.png 1x, https://example.com/logo.png 2x" class="mb-4 h-48 w-full rounded-lg object-cover" src="https://example.com/logo.png">
              <h2 class="text-lg font-semibold">Example Company (example.com) </h2>
              <p class="text-neutral-600">A sample description</p>
              <p class="text-neutral-500"> Status: Approved</p>
              <p class="text-neutral-500"> Reviewed At: 2023-01-01</p>
              <p class="text-neutral-500"> Review Notes: Looks good</p>
              <p class="text-neutral-500"> Reviewed By: admin@example.com</p>
              <p class="text-neutral-500"> Created At: 2022-12-01</p>
            </a>
          </div>
          <div class="rounded-lg bg-white p-4 shadow-md dark:bg-neutral-800">
            <!--v-if--><a href="/users/submit/company?id=2">
              <!--v-if-->
              <h2 class="text-lg font-semibold">Test Company (test.com) </h2>
              <!--v-if-->
              <p class="text-neutral-500"> Status: Pending</p>
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <p class="text-neutral-500"> Created At: 2022-11-15</p>
            </a>
          </div>
        </div>
      </div>
    </main>
  </div>
  <!--v-if-->
</div>"
`;
